### PR TITLE
go_to_line: Avoid byte slicing crash in short status labels

### LINKED ISSUE #40962: MacOS: zed crash
### Description
```

Thread 0 Crashed:: main
0   libsystem_kernel.dylib        	       0x18304a5b0 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x183084888 pthread_kill + 296
2   libsystem_c.dylib             	       0x182f8a898 __abort + 132
3   libsystem_c.dylib             	       0x182f8a814 abort + 136
4   zed                           	       0x10e60a0a4 std::sys::pal::unix::abort_internal::h1edcc850f5dec78e + 12
5   zed                           	       0x10e609208 std::process::abort::hffd6db68ff0662a6 + 12
6   zed                           	       0x10514caac _RNvCsjE14UjDquw0_7crashes10panic_hook + 1380
7   zed                           	       0x10afbf2b0 std::panicking::rust_panic_with_hook::h557a23724a5de839 + 604
8   zed                           	       0x10afbee98 std::panicking::begin_panic_handler::_$u7b$$u7b$closure$u7d$$u7d$::h269cace6208fef05 + 144
9   zed                           	       0x10afbb854 std::sys::backtrace::__rust_end_short_backtrace::h5be0da278f3aaec7 + 12
10  zed                           	       0x10afbeb9c _RNvCsj4CZ6flxxfE_7___rustc17rust_begin_unwind + 32
11  zed                           	       0x10e444318 core::panicking::panic_fmt::h477ff48eff31ffa4 + 32
12  zed                           	       0x104fe502c core::str::slice_error_fail_rt::h224a7c205f145bbe + 480
13  zed                           	       0x10e444738 core::str::slice_error_fail::h5926aa77a149ed5b + 12
14  zed                           	       0x106fc31b0 _RINvMs5_Cs691NJQMz3Qg_4ropeNtB6_6Cursor7summaryNtB6_11TextSummaryECsegUch3txUiq_10go_to_line + 520
15  zed                           	       0x106fcce80 _RINvMse_Cs3O0PTt98SMv_12multi_bufferNtB6_19MultiBufferSnapshot37text_summary_for_excerpt_offset_rangeNtCs691NJQMz3Qg_4rope11TextSummaryECsegUch3txUiq_10go_to_line + 560
16  zed                           	       0x106fcbfd4 _RINvMse_Cs3O0PTt98SMv_12multi_bufferNtB6_19MultiBufferSnapshot22text_summary_for_rangeNtCs691NJQMz3Qg_4rope11TextSummaryNtNtB1o_5point5PointECsegUch3txUiq_10go_to_line + 732
17  zed                           	       0x107012a98 _RINCNCNvMs_NtCsegUch3txUiq_10go_to_line15cursor_positionNtB9_14CursorPosition15update_position00lEBb_ + 4320
18  zed                           	       0x106fe9154 _RNCINvMs5_NtCs3ZRtJULLJf7_4gpui3appNtB8_3App5spawnNCINvMsm_NtBa_6windowNtBV_6Window5spawnNCINvMs0_NtB8_7contextINtB1y_7ContextNtNtCsegUch3txUiq_10go_to_line15cursor_position14CursorPositionE8spawn_inNCNvMs_B22_B20_15update_position0uE0uE0uE0B24_ + 508
19  zed                           	       0x106ff2e24 _RNvMs1_NtCscpu8Xf6C1IL_10async_task3rawINtB5_7RawTaskINtNvNtCs3ZRtJULLJf7_4gpui8executor32spawn_local_with_source_location7CheckedINtNtCs2CqrNpCrSUr_4core3pin3PinINtNtCs1AhsD0INetT_5alloc5boxed3BoxDNtNtNtB29_6future6future6Futurep6OutputuEL_EEEuNCINvNvMs4_BU_NtBU_18ForegroundExecutor5spawn5inneruE0uE3runCsegUch3txUiq_10go_to_line + 652
20  zed                           	       0x1071a68c8 _RNvNtNtNtCs3ZRtJULLJf7_4gpui8platform3mac10dispatcher10trampoline + 20
21  libdispatch.dylib             	       0x182ee2ac4 _dispatch_client_callout + 16
22  libdispatch.dylib             	       0x182f000e4 _dispatch_main_queue_drain.cold.5 + 812
23  libdispatch.dylib             	       0x182ed7f48 _dispatch_main_queue_drain + 180
24  libdispatch.dylib             	       0x182ed7e84 _dispatch_main_queue_callback_4CF + 44
25  CoreFoundation                	       0x183150098 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
26  CoreFoundation                	       0x1831228cc __CFRunLoopRun + 1944
27  CoreFoundation                	       0x1831e0898 _CFRunLoopRunSpecificWithOptions + 532
28  HIToolbox                     	       0x18fb1f730 RunCurrentEventLoopInMode + 316
29  HIToolbox                     	       0x18fb229d0 ReceiveNextEventCommon + 488
30  HIToolbox                     	       0x18fcac1f4 _BlockUntilNextEventMatchingListInMode + 48
31  AppKit                        	       0x1879fa25c _DPSBlockUntilNextEventMatchingListInMode + 236
32  AppKit                        	       0x187510edc _DPSNextEvent + 588
33  AppKit                        	       0x187f63958 -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
34  AppKit                        	       0x187f63664 -[NSApplication(NSEventRouting) nextEventMatchingMask:untilDate:inMode:dequeue:] + 72
35  AppKit                        	       0x187509720 -[NSApplication run] + 368
36  zed                           	       0x1071b4e48 _RNvXs0_NtNtNtCs3ZRtJULLJf7_4gpui8platform3mac8platformNtB5_11MacPlatformNtB9_8Platform3run + 704
37  zed                           	       0x10cbca2a8 _RINvMs1_NtCs3ZRtJULLJf7_4gpui3appNtB6_11Application3runNCNvCs5nDCMQKJDeY_3zed4mains4_0EBV_ + 208
38  zed                           	       0x10d09e18c _RNvCs5nDCMQKJDeY_3zed4main + 5364
39  zed                           	       0x10d124da0 _RINvNtNtCs4N3M0UGf3SB_3std3sys9backtrace28___rust_begin_short_backtraceFEuuECs2pcRUmvhNgu_3zed + 12
40  zed                           	       0x10d124d6c _RNCINvNtCs4N3M0UGf3SB_3std2rt10lang_startuE0Cs2pcRUmvhNgu_3zed + 16
41  zed                           	       0x10afad8e8 std::rt::lang_start_internal::hdb28e94b6865fa11 + 924
42  zed                           	       0x10d124de0 main + 52
43  dyld                          	       0x182cc5d54 start + 7184

Thread 1:: async-io
0   libsystem_kernel.dylib        	       0x183047f30 kevent + 8
1   zed                           	       0x108d00a48 _RNvMs_Cs2FvZQRXzp1p_7pollingNtB4_6Poller9wait_impl + 688
2   zed                           	       0x103e81e68 _RNvMs_NtCs5HiuWjI9q8B_8async_io7reactorNtB4_11ReactorLock5react + 344
3   zed                           	       0x103e7c8a0 _RINvNtNtCs4N3M0UGf3SB_3std3sys9backtrace28___rust_begin_short_backtraceNCNCNvNtCs5HiuWjI9q8B_8async_io6driver8unparker00uEB1f_ + 300
4   zed                           	       0x103e7d168 _RNSNvYNCINvMNtCs4N3M0UGf3SB_3std6threadNtBa_7Builder16spawn_unchecked_NCNCNvNtCs5HiuWjI9q8B_8async_io6driver8unparker00uEs_0INtNtNtCs2CqrNpCrSUr_4core3ops8function6FnOnceuE9call_once6vtableB1e_ + 420
5   zed                           	       0x10afc16c0 std::sys::pal::unix::thread::Thread::new::thread_start::h87df50f049a92661 + 52
6   libsystem_pthread.dylib       	       0x183084c08 _pthread_start + 136
7   libsystem_pthread.dylib       	       0x18307fba8 thread_start + 8
...
```

### Zed Version and System Specs

Zed 0.209.5 
68bb196071e350d3d942ab207e2a70829a512cec

### PR DESCRIPTION
**Pull request by cppcoffee**
Closes https://github.com/zed-industries/zed/issues/40962

Release Notes:

- Fixed avoid byte slicing crash in short status labels

### CODE REVIEW

**Review by parinya-ao — DISMISSED** [review]
LGTM

**Review by SomeoneToIgnore — COMMENTED** [review]
Any chance we can have a test for this?

**Comment by cppcoffee** [comment]
Hi @SomeoneToIgnore , `test_first_char_extraction_utf8_safety` has already tested the fixed core logic. However, the current test cases do not call the interface. Should this test be retained?

**Review by SomeoneToIgnore — COMMENTED** [review]
> `test_first_char_extraction_utf8_safety` has already tested the fixed core logic

Whatever it tested, it was not panicking, right?
My point is to recreate a panic that we fix in a separate test: otherwise it might come back later, because nothing ensures it's not broken.

Thank you for adding the new test, but this test does not fail on `main` for me — which means that we're not recreating a panic that we fix yet.

**Comment by cppcoffee** [comment]
In `cursor_position.rs`, the short-format branch performs `&name[..1]` on the label, which only takes the first byte. When the label is a multi-byte character (like "行"), this slice lands in the middle of a UTF-8 sequence, triggering `core::str::slice_error_fail` and causing a panic.

Introduce the `short_label` helper function to safely truncate short labels at character boundaries, completely eliminating UTF-8 slice panics in `write_position`.

**Comment by SomeoneToIgnore** [comment]
Those are great pieces of context — thank you for explaining the issue, this is usually very hepful to have in the PR initially.
I am insisting on having the test here because it seems to be relatively simple to reproduce, right?
All it needs is a specific text in the buffer + certain input into the go to line modal.

Yet, I have to repeat myself: 

> but this test does not fail on main for me — which means that we're not recreating a panic that we fix yet.

I've taken this PR's diff and restored all the logic that got fixed in this PR:

image

with this diff

```diff
diff --git a/crates/go_to_line/src/cursor_position.rs b/crates/go_to_line/src/cursor_position.rs
index 5c10537e28..fc730c7881 100644
--- a/crates/go_to_line/src/cursor_position.rs
+++ b/crates/go_to_line/src/cursor_position.rs
@@ -162,6 +162,18 @@ impl CursorPosition {
             // Do not write out anything if we have just one empty selection.
             return;
         }
+        self.write_position_with_labels(
+            text,
+            cx,
+            LabelTexts {
+                line: "line",
+                selection: "selection",
+                character: "character",
+            },
+        );
+    }
+
+    fn write_position_with_labels(&self, text: &mut String, cx: &App, labels: LabelTexts<'_>) {
         let SelectionStats {
             lines,
             characters,
@@ -199,12 +211,25 @@ impl CursorPosition {
         &self.selected_count
     }
 
+    fn short_label(name: &str) -> &str {
+        match name.char_indices().nth(1) {
+            Some((index, _)) => &name[..index],
+            None => name,
+        }
+    }
+
     #[cfg(test)]
     pub(crate) fn position(&self) -> Option<UserCaretPosition> {
         self.position
     }
 }
 
+struct LabelTexts<'a> {
+    line: &'a str,
+    selection: &'a str,
+    character: &'a str,
+}
+
 impl Render for CursorPosition {
     fn render(&mut self, _window: &mut Window, cx: &mut Context<Self>) -> impl IntoElement {
         if !StatusBarSettings::get_global(cx).cursor_position_button {
@@ -308,3 +333,66 @@ impl Settings for LineIndicatorFormat {
         content.line_indicator_format.unwrap().into()
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use gpui::{AppContext, TestAppContext};
+    use project::{FakeFs, Project};
+    use serde_json::json;
+    use std::sync::Arc;
+    use util::path;
+    use workspace::{AppState, Workspace};
+    fn init_test(cx: &mut TestAppContext) -> Arc<AppState> {
+        cx.update(|cx| {
+            let state = AppState::test(cx);
+            language::init(cx);
+            crate::init(cx);
+            editor::init(cx);
+            workspace::init_settings(cx);
+            Project::init_settings(cx);
+            state
+        })
+    }
+
+    #[gpui::test]
+    async fn write_position_handles_multibyte_short_labels(cx: &mut TestAppContext) {
+        init_test(cx);
+        let fs = FakeFs::new(cx.executor());
+        fs.insert_tree(path!("/dir"), json!({ "file.txt": "" }))
+            .await;
+        let project = Project::test(fs, [path!("/dir").as_ref()], cx).await;
+        let (workspace, window_cx) =
+            cx.add_window_view(|window, cx| Workspace::test_new(project.clone(), window, cx));
+        let cursor_position = workspace.update(&mut *window_cx, |workspace, cx| {
+            cx.new(|_| CursorPosition::new(workspace))
+        });
+        cx.update(|cx| {
+            LineIndicatorFormat::override_global(LineIndicatorFormat::Short, cx);
+        });
+        cursor_position.update(cx, |cursor_position, cx| {
+            cursor_position.selected_count = SelectionStats {
+                lines: 2,
+                characters: 3,
+                selections: 2,
+            };
+            let mut text = "1:1".to_string();
+            cursor_position.write_position_with_labels(
+                &mut text,
+                cx,
+                LabelTexts {
+                    line: "行",
+                    selection: "选区",
+                    character: "字符",
+                },
+            );
+            let expected = format!(
+                "1:1 (2 {}, 2 {}, 3 {})",
+                CursorPosition::short_label("选区"),
+                CursorPosition::short_label("行"),
+                CursorPosition::short_label("字符"),
+            );
+            assert_eq!(text, expected);
+        });
+    }
+}
```

I'm getting an assertion failure, not a panic that's mentioned in the issue:

```
thread 'cursor_position::tests::write_position_handles_multibyte_short_labels' panicked at crates/go_to_line/src/cursor_position.rs:395:13:
assertion `left == right` failed
  left: "1:1 (2 s, 2 l, 3 c)"
 right: "1:1 (2 选, 2 行, 3 字)"
stack backtrace:
```

Am I doing something wrong, or the test does not really reproduce the panic?

**Comment by mikayla-maki** [comment]
Sorry, this doesn't actually seem to be a bug, nor is it related to the crash reported.

Thanks for your submission!

**Comment by cppcoffee** [comment]
This is because when you "copied the old logic," you didn't actually restore that single-byte truncation of `write_position` back to the original `&name[..1]`. In the new implementation, `write_position` has been split into `write_position_with_labels` which calls `CursorPosition::short_label`, so even if you add back the `label` struct and tests, the code still safely truncates at character boundaries—the test can only reach the assertion, see the English abbreviation, and fail, without triggering a UTF-8 slice panic.

To reproduce the crash, change this line:

  let name = if is_short_format {
      Self::short_label(name)
  } else {
      name
  };

back to the old logic:

  let name = if is_short_format { &name[..1] } else { name };

Then run the unit tests; this will now panic directly ("byte index 1 is not a char boundary"), just like the crash report, proving that the test does indeed cover the path that causes the crash. Then, restore the fix logic, and the test will pass again.

截屏2025-10-25 17 01 54


image

**Comment by SomeoneToIgnore** [comment]
> This is because when you "copied the old logic," you didn't actually restore that single-byte truncation of write_position back to the original `&name[..1]`

Yes I did, did not you see that in the diff provided here:

https://github.com/zed-industries/zed/pull/40963#issuecomment-3437973776

?

> let name = if is_short_format { &name[..1] } else { name };

is unchanged there.

Can you provide me a diff that I can `git apply` on `main` and reproduce the panic?

**Comment by cppcoffee** [comment]
You're right, I've analyzed it and it was indeed my mistake. I'm sorry for the trouble.
